src/platform_impl/android/mod.rs

```diff
@@ -216,383 +189,593 @@ fn ndk_keycode_to_virtualkeycode(keycode: Keycode) -> Option<event::VirtualKeyCo
-                                            }
-                                        }
-                                    }
-                                    InputEvent::KeyEvent(key) => {
-                                        let state = match key.action() {
-                                            KeyAction::Down => event::ElementState::Pressed,
-                                            KeyAction::Up => event::ElementState::Released,
-                                            _ => event::ElementState::Released,
-                                        };
-                                        #[allow(deprecated)]
-                                        let event = event::Event::WindowEvent {
-                                            window_id,
-                                            event: event::WindowEvent::KeyboardInput {
-                                                device_id,
-                                                input: event::KeyboardInput {
-                                                    scancode: key.scan_code() as u32,
-                                                    state,
-                                                    virtual_keycode: ndk_keycode_to_virtualkeycode(
-                                                        key.key_code(),
-                                                    ),
-                                                    modifiers: event::ModifiersState::default(),
-                                                },
-                                                is_synthetic: false,
-                                            },
-                                        };
-                                        call_event_handler!(
-                                            event_handler,
-                                            self.window_target(),
-                                            control_flow,
-                                            event
-                                        );
-                                    }
-                                };
-                                input_queue.finish_event(event, handled);
-                            }
-                        }
-                    }
                 }
-                Some(EventSource::User) => {
-                    // try_recv only errors when empty (expected) or disconnect. But because Self
-                    // contains a Sender it will never disconnect, so no error handling need.
-                    while let Ok(event) = self.user_events_receiver.try_recv() {
-                        call_event_handler!(
-                            event_handler,
-                            self.window_target(),
-                            control_flow,
-                            event::Event::UserEvent(event)
-                        );
-                    }
+                InputEvent::KeyEvent(key) => {
```

Comment:
`android-activity` seems to treat all events as handled by the upper layer similar to what `winit` is currently doing. There is an open issue to declare in particular hardware keys as not-handled to allow volume control. We would probably some way to set the `handled` flag when finishing the event.

Reply:
Definitely agree.

Reply:
Right, technically the NDK InputQueue API, used with `NativeActivity` supports the notion that an application might not handle a key event in case there might be a fallback interpretation but yeah currently the `android-activity` API doesn't expose that notion.

One reason for that was that it's not something that's supported with `GameActivity`, but yeah it could be reasonable to go the other way and allow apps to report events as 'handled' or not and just ignore that with `GameActivity`.

This will require a semver bump, but should be a fairly simple change.

Reply:
Better use a larger rework to solve these issues from the get-go instead of having to rework/refactor again

#1919 / #2432 both describe this issue.

> This will require a semver bump, but should be a fairly simple change.

Shouldn't your crate still be under some alpha/dev notion while working our the final API details within this PR?

Reply:
> > This updates the Android backend to use the [android_activity crate](https://github.com/rib/android-activity) instead
> > of [ndk-glue](https://github.com/rust-windowing/android-ndk-rs/tree/master/ndk-glue). This solves a few issues:
> 
> I have previously requested to keep `ndk-glue` as a pure-Rust glue implementation atop `NativeActivity` and even offered to finalize a rework branch to match the new API (effectively redoing some of the code that currently lives in `winit`, solving exactly the issues that were reported later) as long as we could set up a shared place to define this interface (or toggle it purely at compile time) yet nothing seems to have been done in this regard, making this an instant no-go from me.
> 
> Thanks @msiglreith for making me aware, as I hadn't seen/expected this to make it into a PR yet.

I recall we had some discussion around wanting a pure rust glue layer which I agree would be nice to have eventually but I think I tried to say I saw that as an implementation detail that we can hopefully tackle separately. I think we are hopefully also in a better position with `android-activity` to consider incrementally re-writing the `android_native_app_glue` code - which is about 650 lines of C code.

I tried to say that I believe that the `android_native_app_glue` was a good design / foundation in the short term which has been battle tested over 10 years (for example handling synchronization with the JVM in a way that is abstracted by the glue layer instead of making the application/middleware on top be involved).

Even though the `NativeActivity` backend compiles a small amount C code, that's transparent to the public API here, so I would hope we don't have to be blocked on needing to rewrite that code in Rust right now if it doesn't really fix something?

In terms of the comments about offering to finalize a rework branch to match the new API I'm not sure what this is referring too, so sorry if we had different ideas in some earlier discussion. I tried to look over https://github.com/rust-windowing/winit/issues/2293 but I don't see this. I see these initial questions from you, which might be what you're referring to?:
> - Where does that API live?
> - Who'll write it? Who'll review and sign off on it?
> - Who'll implement/uprev ndk-glue for it?
> - Who'll make winit use the new API?

and after my first pass at implementing `android-activity` I made a proposal and specifically tried to propose an answer for each of these questions: https://github.com/rust-windowing/winit/issues/2293#issuecomment-1173216806

At this point `android-activity` is intended to be able to support a common API for multiple `Activity` classes, starting with `NativeActivity` (can eventually be pure rust) and `GameActivity` and this was arrived at in an incremental way after my first experiments with just supporting GameActivity.

In particular, in [July](https://github.com/rust-windowing/winit/issues/2293#issuecomment-1173216806) I answered your question "Who'll make winit use the new API" with:
> I think the branch needs to be rebased and re-iterated to tidy a few things up but hopefully this is a useful starting point: https://github.com/rib/winit/tree/android-activity

and I didn't see that there was any objection to my proposal or push back - you even sounded positive about some of the issues that I had tackled with its design, such as removing all global static state?

Reply:
I've update the API for `input_events` to make it possible to convey whether an application really handled the event or not so it does not let the InputQueue handle fallbacks: https://github.com/rib/android-activity/pull/32

I'll update this branch/PR once that lands, and after making a corresponding 0.4 release.

Thanks for catching this. I realized I'd left a comment about potentially being able to support fallbacks, potentially higher in the stack for consistency across backends (maybe I was thinking Winit at the time) but having the feedback at this level makes sense to have as a base line.

Reply:
> Shouldn't your crate still be under some alpha/dev notion while working our the final API details within this PR?

I've been dog-fooding the android-activity crate and the winit backend (mainly based on egui usage) for some time now and it's been through a few iterations with an API that I was starting to feel was quite stable / maintainable and so I was actually expecting that the 0.3 release was going to be fine (modulo bug fixes for something that wouldn't affect the API).

Of course, even with the best laid plans we ended up finding a good reason to bump the API here so yeah in retrospect if I'd have used a pre-release tag we could still target 0.3 but bumping to 0.4 isn't really a problem.

Since the intention of the PR is to merge the new backend into master my preference would be for Winit to depend on a stable release of `android-activity`.

Question now is whether I'm optimistic and assume 0.4 should be good to go API-wise, or maybe I'll publish a pre-release this time :)

Reply:
okey, I've published a 0.4.0-beta with the `input_events` API change to allow apps to report events as handled/unhandled.

I've updated the branch for this PR to report all events as `Handled` for now